core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -392,6 +397,8 @@ public CompletableFuture<List<AcquiredRecords>> acquire(String memberId, FetchPa
                 log.trace("There exists another batch which needs to be acquired as well");
                 result.add(acquireNewBatchRecords(memberId, subMap.lastEntry().getValue().lastOffset() + 1,
                     lastBatch.lastOffset(), lastBatch.nextOffset()));
+                applyAcquisitionLock(memberId, subMap.lastEntry().getValue().lastOffset() + 1, lastBatch.lastOffset(),
+                        cachedState.floorEntry(subMap.lastEntry().getValue().lastOffset() + 1));
```

Comment:
Rather than calling in loop why not to keep the API of `applyAcquisitionLock` simple as in accept List<AcquiredRecords>? That will make the code simple to read as well, and potentially bug free, as we only need to call `applyAcquisitionLock` whenever future is completed hence at 2 places in the `acquire` code. wdyt?

Reply:
that would involve us going through more cachedState entries than what might be required. Hence, it will make the performance slower, right?

Reply:
May be you are right and if you explain in some detail then that will be helpful. Essentially you just need to put the batch/subset on timer which can be derived from the result of acquired records. So you would need one iteration on the result list. I am not sure how that would impact the cachedstate entries.

Reply:
So, the result of acquired records can correspond to multiple cachedState entries. If I only pass in the baseOffset and lastOffset to the timer task, there will be additional complexity to identify which cachedEntries does this list of offsets belong to. Since, we are anyways updating the cachedState entries while acquiring records, it made sense we pass the entry object to the timer task to increase the efficiency.

Reply:
passing only baseOffset and lastOffset in the latest implementation to support cancel timeout tasks functionality as well.